examples-gallery/two-pendulum-cart-variable-gallery.py

```diff
@@ -0,0 +1,311 @@
+h = sm.symbols('h')          # the variable time element
+
+state_symbols = tuple((*q_ind, *u_ind))
+constant_symbols = (lx, m1, m2, m3, g, iZZ1, iZZ2)
+specified_symbols = (F,)
+
+target_angle = np.pi /2.0                     
+num_nodes = 250     
+
+duration = (num_nodes - 1) * h
+interval_value = h
+
+# Specify the known system parameters.
+par_map = OrderedDict()
+par_map[lx]   = 2.0
+par_map[m1]  = 1.0
+par_map[m2]  = 1.0
+par_map[m3]  = 1.0
+par_map[g]   = 9.81
+par_map[iZZ1] = 2.0
+par_map[iZZ2] = 2.0
+
+def obj(free):
+    """Minimize h, the time interval between nodes."""
+    return free[-1]
+
+def obj_grad(free):
+    grad = np.zeros_like(free)
+    grad[-1] = 1.
+    return grad
+
+# %%
+# set the initial and final states to form the instance constraints.
+# initially, the pendulum should be hanging straight down, without mvoing
+# and finally, the pendulum should be upright, without moving.
+initial_state_constraints = {
+                            q1: 0.,  
+                            q2: -np.pi/2., 
+                            q3: -np.pi/2., 
+                            u1: 0.,
+                            u2: 0., 
+                            u3: 0.
+                            }
+
+final_state_constraints  = {
+                           q2: target_angle, 
+                           q3: target_angle, 
+                           u1: 0., 
+                           u2: 0., 
+                           u3: 0.
+                           }
+
+instance_constraints = (
+) + tuple(
+    xi.subs({t: 0}) - xi_val for xi, xi_val in initial_state_constraints.items()
+) + tuple(
+    xi.subs({t: duration}) - xi_val for xi, xi_val in final_state_constraints.items()
+)
+# bounding h > 0 herlps avoid 'solutions' with h < 0.
+bounds = {F: (-150., 150.), q1: (-5, 5), h: (1.e-5, 1.0)} 
+
+# %%
+# Create an optimization problem and solve it.
+prob = Problem(obj, 
+               obj_grad, 
+               EOM, 
+               state_symbols, 
+               num_nodes, 
+               interval_value,
+               known_parameter_map=par_map,
+               instance_constraints=instance_constraints,
+               time_symbol=t,
+               bounds=bounds)
+
+# Initial guess.
+initial_guess =  np.zeros(prob.num_free)
+initial_guess[-1] = 0.01
+
+# allows to change the number of iterations.
+# standard is 3000
+prob.add_option('max_iter', 6000)
+
+# Find the optimal solution.
+# iterating and using the solution of iteration i
+# as initial guess for iteration i+1 sometimes helps.
+for i in range(2):
+    solution, info = prob.solve(initial_guess)
+    initial_guess = solution
+    
+    print(f'{i + 1}st iteration: value of objective function is {obj(solution):.2f}')
+    prob.plot_objective_value()
+    print('message from optimizer:', info['status_msg'])
+    print('Iterations needed',len(prob.obj_value))
+    print(f"objective value {info['obj_val']:.3e}")
+    print('h value:', solution[-1])
+    print('\n')
```

Comment:
This is likely not appropriate for a example if we have to do multiple solves. Best to give a better initial guess to get the # iterations down to <500 for an optimal solution.